rules/no-accessor-recursion.js

```diff
@@ -0,0 +1,97 @@
+const MESSAGE_ID_ERROR = 'no-accessor-recursion/error';
+const messages = {
+	[MESSAGE_ID_ERROR]: 'Disallow recursive access to `this` within getters and setters.',
+};
+
+/** @param {import('eslint').Scope.Scope} scope */
+const isArrowFunctionScope = scope => scope.type === 'function' && scope.block.type === 'ArrowFunctionExpression';
+
+/**
+Get the closest non-arrow function scope.
+
+@param {import('eslint').SourceCode} sourceCode
+@param {import('estree').Node} node
+*/
+const getClosestFunctionScope = (sourceCode, node) => {
+	let scope = sourceCode.getScope(node);
+	while (scope.type !== 'function' || isArrowFunctionScope(scope)) {
+		if (scope.upper === null) {
+			return scope;
+		}
+
+		scope = scope.upper;
+	}
+
+	return scope;
+};
+
+const isDotNotationAccess = node => node.type === 'MemberExpression' && !node.computed && node.property.type === 'Identifier';
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const {sourceCode} = context;
+	const functionExpressionStack = [];
+
+	return {
+		...Object.fromEntries(
+			['Property', 'MethodDefinition'].flatMap(nodeName =>
+				['get', 'set'].flatMap(kind => [
+					[`${nodeName}[kind="${kind}"]`, node => {
+						functionExpressionStack.push(node);
+					}],
+					[`${nodeName}[kind="${kind}"]:exit`, () => {
+						functionExpressionStack.pop();
+					}],
+				]),
+			),
+		),
+		/** @param {import('estree').ThisExpression} node */
+		ThisExpression(node) {
+			/** @type {import('estree').Property | import('estree').MethodDefinition} */
+			const property = functionExpressionStack.at(-1);
+
+			if (!property) {
+				return;
+			}
+
+			const scope = getClosestFunctionScope(sourceCode, node);
```

Comment:
Nice refactoring.

Now we have implemented this functionality, do we still need the functionExpressionStack? we can check `scope.block.parent` directly instead?

Reply:
Yes, this is indeed better.